## Introduction to Data Science

#### University of Redlands - DATA 101
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data101.joannabieri.com](https://joannabieri.com/data101.html)

---------------------------------------
# Homework Day 7
---------------------------------------

GOALS:

1. Answer all the questions from the lecture - we will use the scientists data and the grocery data.
2. Analyze data about College majors - your first EDA! (Think of this as a practice exam.)

----------------------------------------------------------

This homework has **9 Questions** and **A Practice Exam**

NOTE:
The practice exam is intended to give you an idea of what an exam in this class might feel like. Remember you will be expected to work on the exam **independently** and submit your work before we get together to work on the exam in class. **15\% of your Exam grade will come from what you can do independently** so start practicing now and get help if you need it!.

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

from itables import show

## Lecture Questions

First we will import the data about female scientists.

In [2]:
file1 = 'https://joannabieri.com/introdatascience/data/dates.csv'
DF_dates = pd.read_csv(file1)
file2 = 'https://joannabieri.com/introdatascience/data/professions.csv'
DF_professions = pd.read_csv(file2)
file3 = 'https://joannabieri.com/introdatascience/data/works.csv'
DF_works = pd.read_csv(file3)
show(DF_dates)
show(DF_professions)
show(DF_works)

**Q1** Are each of these data sets Tidy?

**Q2** Do they all contain the same number of observations?

**Q3** What are the five variables?

**Q4** What variable do they have in common - what can we join **on**

**(Click Here)**

1) The data sets are each tidy, the last one is arguable though, as it is both a list of accomplishments as well as a list of several different subjects
2) No, the first data set has 3 observations while the others have 2
3) The 5 variables are name, birth_year, death_year, profession, and known_for
4) All of these data sets have the common variable of name.

## Combining Three Data Sets

You need to think about the order and whether or not you want to keep the maximum number of names or only keep names that are in all the data frames.

In this case we will try to keep the maximal amount of data - the maximum number of names.

1. Combine the professions and works data
2. Then add the dates data

In [17]:
# 1. Combine the professions and works data
DF_scientists = pd.merge(DF_professions,DF_works,on='name',how='left')
DF_scientists
# 2. Then add the dates data
DF_scientists = pd.merge(DF_scientists,DF_dates,on='name',how='left')
DF_scientists

name                          profession  \
0        Ada Lovelace                       Mathematician   
1         Marie Curie               Physicist and Chemist   
2        Janaki Ammal                            Botanist   
3     Chien-Shiung Wu                           Physicist   
4   Katherine Johnson                       Mathematician   
5   Rosalind Franklin                             Chemist   
6          Vera Rubin                          Astronomer   
7         Gladys West                       Mathematician   
8  Flossie Wong-Staal  Virologist and Molecular Biologist   
9     Jennifer Doudna                          Biochemist   

                                           known_for  birth_year  death_year  
0                           first computer algorithm         NaN         NaN  
1  theory of radioactivity,  discovery of element...         NaN         NaN  
2            hybrid species, biodiversity protection      1897.0      1984.0  
3  confim and refine theory of radioactive beta d...      1912.0      1997.0  
4  calculations of orbital mechanics critical to ...      1918.0      2020.0  
5                                                NaN      1920.0      1958.0  
6                           existence of dark matter      1928.0      2016.0  
7  mathematical modeling of the shape of the Eart...      1930.0         NaN  
8  first scientist to clone HIV and create a map ...      1947.0         NaN  
9  one of the primary developers of CRISPR, a gro...      1964.0         NaN

**Q5** What, if anything, would change if you switched to how='right'?

**(Click Here)**
The ones with nan statistics for both birth_year and death_year are no longer present, as it looks at the chart on the right first I believe.

## Combining Two Data Sets - You Try

**Q6** Write code that would combine the professions data and the dates data, but drop any names that don't appear in both.

In [19]:
# Your code here
DF_scientists = pd.merge(DF_professions, DF_dates, on = 'name', how='right')
DF_scientists


name                          profession  birth_year  \
0        Janaki Ammal                            Botanist        1897   
1     Chien-Shiung Wu                           Physicist        1912   
2   Katherine Johnson                       Mathematician        1918   
3   Rosalind Franklin                             Chemist        1920   
4          Vera Rubin                          Astronomer        1928   
5         Gladys West                       Mathematician        1930   
6  Flossie Wong-Staal  Virologist and Molecular Biologist        1947   
7     Jennifer Doudna                          Biochemist        1964   

   death_year  
0      1984.0  
1      1997.0  
2      2020.0  
3      1958.0  
4      2016.0  
5         NaN  
6         NaN  
7         NaN

## Combining Three Data Sets - You Try

**Q7** Try to combine the three data sets together (professions, works, and dates) but instead of doing what we did above, see if you can keep only the names that appeared in all three data sets.

In [23]:
# Your code here
DF_scientists = pd.merge(DF_professions, DF_works, on = 'name', how = 'left')
DF_scientists = pd.merge(DF_scientists, DF_dates, on = 'name', how = 'right')
DF_scientists


name                          profession  \
0        Janaki Ammal                            Botanist   
1     Chien-Shiung Wu                           Physicist   
2   Katherine Johnson                       Mathematician   
3   Rosalind Franklin                             Chemist   
4          Vera Rubin                          Astronomer   
5         Gladys West                       Mathematician   
6  Flossie Wong-Staal  Virologist and Molecular Biologist   
7     Jennifer Doudna                          Biochemist   

                                           known_for  birth_year  death_year  
0            hybrid species, biodiversity protection        1897      1984.0  
1  confim and refine theory of radioactive beta d...        1912      1997.0  
2  calculations of orbital mechanics critical to ...        1918      2020.0  
3                                                NaN        1920      1958.0  
4                           existence of dark matter        1928      2016.0  
5  mathematical modeling of the shape of the Eart...        1930         NaN  
6  first scientist to clone HIV and create a map ...        1947         NaN  
7  one of the primary developers of CRISPR, a gro...        1964         NaN

## Case study - Grocery Sales

Below you will load some data about grocery sales and see how joining or merging data can help us answer questions.

In [24]:
file1 = 'https://joannabieri.com/introdatascience/data/purchases.csv'
DF_purchases = pd.read_csv(file1)
file2 = 'https://joannabieri.com/introdatascience/data/prices.csv'
DF_prices = pd.read_csv(file2)
show(DF_purchases)
show(DF_prices)

## Calculate the total revenue

**Q8** Given the data above find the total revenue.

We will have to join the data frames so that we can see the price of what was sold. So first we look for a common column. In this case both data frames have **item** as a column.

Now, think about how to calculate revenue... we need to add up the total money we made, so we need to know the price of each thing sold.

**PAUSE - see if you can write some of the code for this before looking at the cells below**

In [27]:
DF_items = pd.merge(DF_purchases, DF_prices, on = 'item', how = 'right')
DF_items

customer_id          item  price
0          NaN       avocado   0.50
1          1.0        banana   0.15
2          1.0         bread   1.00
3          1.0          milk   0.80
4          2.0          milk   0.80
5          2.0  toilet paper   3.00

In [28]:
DF_items.price.sum()

6.25

In [9]:
#$6.25

In [10]:
#

**See the lecture notes if you need heelp figuring this out!**

**Q9** Calculate the revenue per customer? Hint - group by the customer id and then apply the sum().

In [29]:
# Your code here
DF_items.groupby('customer_id').sum()
#id 1 bought $1.95, id 2 bought $3.80

item  price
customer_id                         
1.0           bananabreadmilk   1.95
2.0          milktoilet paper   3.80

# Practice Exam - EDA - College Majors

Your solutions should include code written to answer the questions AND words describing the results. For each problem:

1. Write code to get the answers. It is okay to copy and paste from our lecture notes or your homework assignments.
2. Describe what the code output is telling you.
3. Explain any interesting aspects of your results.

------------------------
-------------------------

The first step in the process of turning information into knowledge process is to summarize and describe the raw information - the data.
In this assignment we explore data on college majors and earnings, specifically the data begin the FiveThirtyEight story ["The Economic Guide To Picking A College Major"](https://fivethirtyeight.com/features/the-economic-guide-to-picking-a-college-major/).

These data originally come from the American Community Survey (ACS) 2010-2012 Public Use Microdata Series.
While this is outside the scope of this assignment, if you are curious about how raw data from the ACS were cleaned and prepared, see [the code](https://github.com/fivethirtyeight/data/blob/master/college-majors/college-majors-rscript.R) FiveThirtyEight authors used.

We should also note that there are many considerations that go into picking a major.
Earnings potential and employment prospects are two of them, and they are important, but they don't tell the whole story.
Keep this in mind as you analyze the data.

In [31]:
file_location = 'https://joannabieri.com/introdatascience/data/recent-grads.csv'
DF_raw = pd.read_csv(file_location)

In [32]:
DF_raw

Rank  Major_code                                      Major    Total  \
0       1        2419                      PETROLEUM ENGINEERING   2339.0   
1       2        2416             MINING AND MINERAL ENGINEERING    756.0   
2       3        2415                  METALLURGICAL ENGINEERING    856.0   
3       4        2417  NAVAL ARCHITECTURE AND MARINE ENGINEERING   1258.0   
4       5        2405                       CHEMICAL ENGINEERING  32260.0   
..    ...         ...                                        ...      ...   
168   169        3609                                    ZOOLOGY   8409.0   
169   170        5201                     EDUCATIONAL PSYCHOLOGY   2854.0   
170   171        5202                        CLINICAL PSYCHOLOGY   2838.0   
171   172        5203                      COUNSELING PSYCHOLOGY   4626.0   
172   173        3501                            LIBRARY SCIENCE   1098.0   

         Men    Women            Major_category  ShareWomen  Sample_size  \
0     2057.0    282.0               Engineering    0.120564           36   
1      679.0     77.0               Engineering    0.101852            7   
2      725.0    131.0               Engineering    0.153037            3   
3     1123.0    135.0               Engineering    0.107313           16   
4    21239.0  11021.0               Engineering    0.341631          289   
..       ...      ...                       ...         ...          ...   
168   3050.0   5359.0    Biology & Life Science    0.637293           47   
169    522.0   2332.0  Psychology & Social Work    0.817099            7   
170    568.0   2270.0  Psychology & Social Work    0.799859           13   
171    931.0   3695.0  Psychology & Social Work    0.798746           21   
172    134.0    964.0                 Education    0.877960            2   

     Employed  ...  Part_time  Full_time_year_round  Unemployed  \
0        1976  ...        270                  1207          37   
1         640  ...        170                   388          85   
2         648  ...        133                   340          16   
3         758  ...        150                   692          40   
4       25694  ...       5180                 16697        1672   
..        ...  ...        ...                   ...         ...   
168      6259  ...       2190                  3602         304   
169      2125  ...        572                  1211         148   
170      2101  ...        648                  1293         368   
171      3777  ...        965                  2738         214   
172       742  ...        237                   410          87   

     Unemployment_rate  Median  P25th   P75th  College_jobs  Non_college_jobs  \
0             0.018381  110000  95000  125000          1534               364   
1             0.117241   75000  55000   90000           350               257   
2             0.024096   73000  50000  105000           456               176   
3             0.050125   70000  43000   80000           529               102   
4             0.061098   65000  50000   75000         18314              4440   
..                 ...     ...    ...     ...           ...               ...   
168           0.046320   26000  20000   39000          2771              2947   
169           0.065112   25000  24000   34000          1488               615   
170           0.149048   25000  25000   40000           986               870   
171           0.053621   23400  19200   26000          2403              1245   
172           0.104946   22000  20000   22000           288               338   

     Low_wage_jobs  
0              193  
1               50  
2                0  
3                0  
4              972  
..             ...  
168            743  
169             82  
170            622  
171            308  
172            192  

[173 rows x 21 columns]

### 1. Describe the data that you see here.

* How many variables?
* How man observations?
* Are the variables categorical or numerical?
* Are there any variables that you don't know what they mean? If so [CLICK HERE - to see the Git Repo for the data where the variables are defined](https://github.com/fivethirtyeight/data/tree/master/college-majors)


In [44]:
show(DF_raw)


**(Click Here to Explain your Results)**
21 different variables, 173 different observations
Major and Major_category are categorical variables, the rest are numerical variables.

### 2. Answer some questions about the data

* Which major has the lowest unemployment rate? Show a data frame that answers this question, but only show columns that are useful to the answer.
* Which major has the highest percentage of women? Show a data frame that answers this question, but only show columns that are useful to the answer.
* How do the distributions of median income compare across major categories?
    * First get a value count on what major categories are in the data. What categories are most or least represented?
    * Do some summary statistics using the .group_by() command to see the statistics for each major category.
    * Make a histogram that explores median income for each of the major categories (facet_col). How did you choose your number of bins?
    * Why do we often choose the median, rather than the mean, to describe the typical income of a group of people? Look this up online if you don't know the answer!
    * Answer the question: Which major category has the highest typical (you'll need to decide what this means) median income?

In [50]:
unemployment = ['Major', 'Unemployment_rate']
DF_raw[unemployment].sort_values('Unemployment_rate')


Major  Unemployment_rate
120  EDUCATIONAL ADMINISTRATION AND SUPERVISION           0.000000
73                        MILITARY TECHNOLOGIES           0.000000
83                                       BOTANY           0.000000
52             MATHEMATICS AND COMPUTER SCIENCE           0.000000
112                                SOIL SCIENCE           0.000000
..                                          ...                ...
29                                PUBLIC POLICY           0.128426
170                         CLINICAL PSYCHOLOGY           0.149048
84   COMPUTER NETWORKING AND TELECOMMUNICATIONS           0.151850
89                        PUBLIC ADMINISTRATION           0.159491
5                           NUCLEAR ENGINEERING           0.177226

[173 rows x 2 columns]

In [60]:
women = ['Major', 'Women','ShareWomen']
DF_raw[women].sort_values('ShareWomen',ascending=False)


Major     Women  ShareWomen
164                      EARLY CHILDHOOD EDUCATION   36422.0    0.968954
163  COMMUNICATION DISORDERS SCIENCES AND SERVICES   37054.0    0.967998
51                      MEDICAL ASSISTING SERVICES   10320.0    0.927807
138                           ELEMENTARY EDUCATION  157833.0    0.923745
150                   FAMILY AND CONSUMER SCIENCES   52835.0    0.910933
..                                             ...       ...         ...
1                   MINING AND MINERAL ENGINEERING      77.0    0.101852
26                           CONSTRUCTION SERVICES    1678.0    0.090713
66     MECHANICAL ENGINEERING RELATED TECHNOLOGIES     371.0    0.077453
73                           MILITARY TECHNOLOGIES       0.0    0.000000
21                                    FOOD SCIENCE       NaN         NaN

[173 rows x 3 columns]

In [70]:

DF_raw['Major_category'].combine.

TypeError: Series.combine() missing 2 required positional arguments: 'other' and 'func'

**(Click Here to Explain your Results)**
Several majors had a 0% unemployment rate, but EARLY CHILDHOOD EDUCATION was the major with the largest percent of women. 

### 3. More complicated questions

* Are all STEM fields the same in terms job opportunities?
    * Assume the following are STEM major categories
    
  ``` ["Biology & Life Science", "Computers & Mathematics", "Engineering","Physical Sciences"]```
  
    * Create a DataFrame (using a mask) so you only look at STEM majors.
    * Which STEM majors have median salaries equal to or less than the median for all majors' median earnings?
* What types of majors do women tend to major in?
    * Create a scatterplot of median income vs. proportion of women in that major, coloured by whether the major is in a STEM field or not.

In [21]:
#

In [22]:
#

In [23]:
#

**(Click Here to Explain your Results)**

### 4. Ask a question of interest to you.

In [24]:
#

**(Click Here to Explain your Results)**